### PR TITLE
Azure Pipelines: Skip tests if build fails, run both tools even if other fails

### LINKED ISSUE #1266: Azure Pipelines: When the build fails, we shouldn't run tests
Currently, if a PR/CI build fails, we attempt to run the tests anyways. This is doomed and provides no useful info.

To fix this, it appears that `azure-pipelines.yml` and `azure-devops/run-build.yml` should `name:` their steps, and use `dependsOn:` to establish dependencies. Then, the [default condition](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/conditions?view=azure-devops&tabs=yaml) of "Only when all previous dependencies have succeeded" should be effective.

I don't immediately know all of the dependencies that we should set up. The STL tests should definitely depend on the STL build. However, I suspect that "Publish Tests" should depend on the STL build, but should run regardless of whether the STL tests succeeded or failed. Similarly, clang-format and validate enforcement should depend on Build Support Tools, but they should run regardless of whether the other fails.

We probably don't need to set up literally every dependency imaginable - just ones that prevent huge wastes of time.

### PR DESCRIPTION
**Pull request by sylveon**
Fixes #1266

This adds basic build dependencies to the CI config. I've only added some basic dependencies to keep it simple, since the rest should eventually cascade to a build failure.

### CODE REVIEW

**Comment by BillyONeal** [comment]
High 5 for getting the best PR number possible.

**Comment by sylveon** [comment]
After making my other PR I just had to grab that number :P

Looks like steps don't support dependsOn, but I might be able to use condition instead, looking through the docs ATM.

**Comment by sylveon** [comment]
Conditions in steps won't really work because you can't verify if a specific step failed or not.

One idea that I have (at least for the validation stage) is to split it in 3 jobs:
- One that builds the tooling
- One that runs clang-format
- One that validates files

So that way, we can specify job dependencies normally. Same idea could be applied to the build & test jobs if each architecture is its own stage, rather than its own job.

**Comment by CaseyCarter** [comment]
Looks like this is Work in Progress? Shout if I'm wrong, or when you push some changes and it's ready for review.

**Comment by sylveon** [comment]
Yeah, but this is CI changes and IIRC draft PRs don't trigger CI, but since my changes are about CI...

**Comment by StephanTLavavej** [comment]
Makes sense - thanks for investigating this, and feel free to continue pushing commits to trigger the CI. (This PR will ultimately save us lots of resources when builds fail!)

**Comment by SuperWig** [comment]
> draft PRs don't trigger CI

Unless that changed recently it should.

**Comment by CaseyCarter** [comment]
> Yeah, but this is CI changes and IIRC draft PRs don't trigger CI, but since my changes are about CI...

That wasn't griping about making it a Draft PR - I can never remember if they trigger CI - just communicating my understanding of its status and choice of stage in https://github.com/microsoft/STL/projects/1. I wanted to ensure we're on the same page so you're not inadvertently left sitting here waiting for maintainer feedback.

**Comment by sylveon** [comment]
@CaseyCarter I believe this is ready for review!

Here's an example of a failing build that skips tests: https://dev.azure.com/vclibs/STL/_build/results?buildId=5335&view=results

I also squashed all my commits because testing CI is p a i n

File: azure-pipelines.yml
Comment (cbezault) [inline azure-pipelines.yml:41]: It seems sort of overkill to publish this as an artifact and then run it after downloading it again. What was wrong with just building and running without this intermediate publishing step?

Comment (sylveon) [reply azure-pipelines.yml:41]: It's split in different jobs now (to allow one to run despite the other having failed), which means they can/will end up running on a different worker which doesn't have access to the build artifacts.

**Review by CaseyCarter — APPROVED** [review]

File: azure-devops/run-build.yml
Comment (sylveon) [reply azure-devops/run-build.yml:8]: Right, according to https://docs.microsoft.com/en-us/azure/devops/pipelines/yaml-schema?view=azure-devops&tabs=schema%2Cparameter-schema#pool.

Comment (CaseyCarter) [reply azure-devops/run-build.yml:8]: Great - and thanks for the doc link!

File: azure-pipelines.yml
Comment (cbezault) [reply azure-pipelines.yml:41]: I don't see why you can't just have both the validate step and the clang-format step run no matter what.

Comment (sylveon) [reply azure-pipelines.yml]: Fair enough, I've reverted to running them both no matter what.

File: azure-pipelines.yml
Comment (CaseyCarter) [inline azure-pipelines.yml:44]: Won't this run the enforcement script even if building the (necessary) support tools (the previous step) fails?

Comment (sylveon) [reply azure-pipelines.yml:44]: Yes, but it'll just immediately fail due to not finding the tool. Specifying dependencies requires to use jobs, which means we have to bother with artifacts.

Comment (CaseyCarter) [reply azure-pipelines.yml:44]: So _this_ line could as well be the default (`succeeded()`); it's fine for this step to fail if the previous step failed.

Comment (CaseyCarter) [reply azure-pipelines.yml:44]: ...or should we mark both of these steps `continueOnError`? It seems like that would be equivalent since there are no further steps.

Comment (sylveon) [reply azure-pipelines.yml:44]: `continueOnError` does not make the build error, it turns those failures into warnings only

Comment (CaseyCarter) [reply azure-pipelines.yml:44]: That would certainly not be equivalent, then. Thanks for the clarifications.

**Review by cbezault — APPROVED** [review]

**Comment by cbezault** [comment]
Sorry I made you throw away some of your work :(
LGTM now

**Comment by sylveon** [comment]
Haha it's fine I often throw away my work myself.

**Review by StephanTLavavej — APPROVED** [review]
Thanks for this significant infrastructure improvement - I think it'll improve the experience for contributors *and* decrease load on the VMSS which is awesome.

**Comment by CaseyCarter** [comment]
FYI, I'm about to merge this. If the CI stops working, I will likely snap my fingers Thanatos-style and destroy a significant portion of the universe.

**Comment by CaseyCarter** [comment]
Thanks for the contribution! (I'm 100% behind anyone-but-me enjoying the "edit-submit-wait an hour for CI" loop!)

**Pull request merged**
